test_regress/t/t_xml_flat.out

```diff
@@ -11,10 +11,10 @@
     <file id="d" filename="t/t_xml_first.v" language="1800-2017"/>
   </module_files>
   <cells>
-    <cell fl="d7" loc="d,7,8,7,9" name="TOP" submodname="TOP" hier="TOP"/>
+    <cell fl="d7" loc="d,7,8,7,9" name="root" submodname="root" hier="root"/>
```

Comment:
Hmm. Thought the internal name would still be top?  This suggests to me we should perhaps use $root (which will translate to __024root)?  Then we really are compliant with what one would expect for IEEE?  

But then the filenames will have the 024root.  That is what we do with __024unit currently though....

Not sure right thing here.

Reply:
Intention was for the module to be called "root", and it's instance (top scope) is called "TOP" (this is what I was eluding toward in my question about renaming the top scope) but seems like I changed both. (The top scope is still called TOP though...). Semantically this is `$root` though (and what LinkDot gives you when you refer to `$root`.

Well, I think a `$` is valid in a filename.... But guaranteed trouble via Make secondary expansions or similar, so let's not do that.

Maybe just expand `$` in filenames to `_` so it becomes `{prefix}__root`? Would still clash with a user defined `_root`..

Reply:
Sorry I wasn't clear.  "$root" would be the Verilog name.  In Verilator it would be through the normal quoting Verilator does "__024root", so no problems with make.  A user would need to use "$root" to conflict in Verilog which is of course illegal, hence we'd never need to think about conflicts.

Reply:
I'm ok with that, it's just the oddity of the user having to include __024root.h for access to internals, are we happy to live with that?

Reply:
I think TOP for the cell name and __024root for the module/file name?

Reply:
(as you have a good point)

Reply:
Ok I'm more confused now. It's the module/filename (with the __024 escape) that the user sees, the "TOP" instance name is mostly internal.

Reply:
Another way of fixing this, is to never inline anything into the root scope (i.e.: keep the user specified top levels as cells). Given that often all you have in the wrapper are a clock and reset pins, this might not be too expensive, and then any internals access can be done via the cell pointers in the model class, no need to access the contents of `$root` at all. Thoughts?

Reply:
> this might not be too expensive

In terms of performance that is.

Reply:
I don't think we should change inlining.  I think the proposal for discussion is this:

       <cell fl="d7" loc="d,7,8,7,9" name="TOP" submodname="$root" hier="TOP"/>

That is name is TOP, and Verilog modname is $root which will show up in filenames as __024root.  (It might also be __024root in the XML, not sure which the XML prints.)

Reply:
Note this is in contrast to 

           <cell fl="d7" loc="d,7,8,7,9" name="$root" submodname="$root" hier="$root"/>

which might be more standard but seems disruptive.  Open to debate.

Reply:
(Part of the confusion is a Verilated model might not really end up the root in the larger model, so foo.bar.$root seems odd.

Reply:
> Verilated model might not really end up the root in the larger model#

But isn't that solely determined by what you pass into the model constructor as name? (which currently defaults to "TOP", which I agree we should not change). I wonder whether there is a point in having a `$root` cell in the XML at all, that in itself is a bit odd. My claim is that the user can't actually see the instance name assigned to `$root` unless reading the guts of the generated code, and except maybe in that line on the XML?

Reply:
BTW, that XML line comes from here (as there really isn't a real AstCell for `$root`), so the cell and hier name is arbitrarily chosen to be the same as the module name:
https://github.com/verilator/verilator/blob/2cdf0aae727333148c30425bbb2d9224cfd70fac/src/V3EmitXml.cpp#L323-L329

Reply:
Yes, I think we're agreed the "TOP" used in model construction will remain TOP.  I think the user does see this name though in the generated filenames.  Given that code is the proposal then we use $root for the XML line?

   <cell fl="d7" loc="d,7,8,7,9" name="$root" submodname="$root" hier="$root"/>

Reply:
Further to the story, the cell for the `$root` in the XML only shows up if you use `--flatten` together with `--xml-only`, as we don't create a wrapper in `--xml-only` mode otherwise:
https://github.com/verilator/verilator/blob/2cdf0aae727333148c30425bbb2d9224cfd70fac/src/Verilator.cpp#L201-L206

---

I think I am loosing track of what we are discussing :)

There are at least 5 names involved here:
1. Dynamic name of top level, passed at model instantiation time.
2. Name of `$root`  module (thing that is created in V3LinkLevel::wrapTop. This does not have a corresponding `AstCell` at all). (Emitted in the XML as the `submod=` part)
3. Name of `$root` instance (cell) as emitted in the XML with `--flaten` (the `name=` part)
4. Hierarchy path to `$root` instance (cell) as emitted in the XML with `--flaten` (the `hier=` part)
5. The name of the top level scope (currently "TOP"), which I believe is created here:
https://github.com/verilator/verilator/blob/2cdf0aae727333148c30425bbb2d9224cfd70fac/src/V3Scope.cpp#L95-L110

Point 3 and 4 are a bit of a red herring as they only show up with `--xml-only --flatten`, and can be set arbitrarily either when creating the wrapper in xml only mode, or in EmitXml, as these will not appear anywhere else,

Point 5 is what appears in any flattened variable name, function names, module instance name etc, basically anywhere where the hierarchical scope name is involved in the internals, examples:
- inlined/auxiliary internal variable: `__Vclklast__TOP____Vcellinp__t__clk`
- internal function names: `Vt_cover_line_sc_t___sequent__TOP__t__2`
- Module instances in Syms: `Vt_cover_line_sc_beta          TOP__t__b1`
I again claim that this is only observable in code internals, and not by the public user interface or filenames. This is what I was referring to when I asked if you want me to change this (the top scope name) to `root`, but again, as this is only visible in emitted code internals, it is not important to change, just might make the emitted code internals a bit more readable/consistent, but also if user code pokes around inside, it would need changing, so best leave this alone.

Point 1, the user provided model instance name defaults to "TOP" via the default value to the name constructor argument. This is only observable in trace files and VPI paths. Most importantly, this user given name replaces (or at least should replace) the top scope name (point 5) at runtime in every relevant context that is user observable on purpose. (note there is no VPI scope for `$root`). We agree we should not change the default on this.

This leaves us with point 2, the name of the `$root` module, which used to be "TOP" but that clashes with user defined `module top` which is common. This is what shows up in filenames, and class names. I have changed this to `root` (or `_root` if there is a user defined `module root`, or `__root` if there is also a `module _root`, etc.)

With all that in mind, I propose:
- Leave alone point 1, as agreed.
- Leave alone point 5 (so user code need not be updated too much).
- Point 3 & 4, we have some alternatives
  a. Leave it alone, or more specifically, let it come out as it does with the current code in EmitXml, as this really is just an artifact of using `--flatten` with `--xml-only`, and probably is not very important so long as it matches the emitted singleton module in the XML (which it does)
  b. in `--xml-only --flatten` mode, we revert to calling the `$root` wrapper "TOP".
  c. in `--xml-only --flatten` mode, we create the `$root` wrapper, but then flatten everything into the user defined top levels, rather than into the `$root` wrapper. Semantically this probably is the cleanest, but it would mean flattening multiple user defined top levels would stay as multiple modules, rather than a single one, but it is all a bit arbitrary what the right answer is when "flattening multiple top levels and producing XML output". Note for single user top level, this would work as expected.
- Point 5, we have 2 options:
  a. Leave as implemented in this PR right now (called `root`, or with `_` prefixes when clashing).
  b. Change name to `$root`, and update docs to explain user needs `{prefix}__024root.h` for internals access via the `rootp` in the model class. SV semantics wise this is cleaner, it's just the `_024` escape is odd, but that can be documented.

Would you mind picking your preference from Point 3/4 {a,b,c} and Point 5 {a, b}? I'm happy to do the work for whatever is cleanest. Alternatively, if we are still not talking about the same thing, then please advise :)

Reply:
BTW I'm also a bit puzzled as to the purpose of `--xml-only` with `--flatten`, as I thought `--xml-only` was basically there to generate a simpler machine readable description of the design, so why `--flatten`?

Reply:
Thanks for enumerating that helps the discussion.

> Point 3 & 4, we have some alternatives
> a. Leave it alone, or more specifically, let it come out as it does with the current code in EmitXml, as this really is just an artifact of using --flatten with --xml-only, and probably is not very important so long as it matches the emitted singleton module in the XML (which it does)

Suggest we follow this.

>    Point 5, we have 2 options:
>    b. Change name to $root, and update docs to explain user needs {prefix}__024root.h for internals access via the rootp in the model class. SV semantics wise this is cleaner, it's just the _024 escape is odd, but that can be documented.

Suggest this.

> BTW I'm also a bit puzzled as to the purpose of `--xml-only` with `--flatten`, as I thought `--xml-only` was basically there to generate a simpler machine readable description of the design, so why `--flatten`?

See #2270 - depends on what level the downstream tool wants to operate.  There's probably only a single tool (the requestor's) that needs this, so not terribly important to have it be compatible, so I think we should have whatever falls out be what we do  (thus picked 3.a ).

Reply:
Ok, all done on this one I believe.